In [67]:
import pandas as pd
# Load dataset
data_path = '../data/OnlineRetail.csv'  # Adjust path if needed
df = pd.read_csv(data_path, encoding='latin1')  # Encoding needed for special characters
# Preview data
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [68]:
print(f"Dataset shape: {df.shape}")
print(df.info())

Dataset shape: (541909, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB
None


In [69]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [70]:
df_clean = df.dropna(subset=['CustomerID']).copy()

Many records with missing Customer ID, which makes it inefficient for analysis purposes based on the customer. We remove NaN rows of CustomerID column.

In [71]:
df_clean['InvoiceDate'] = pd.to_datetime(df_clean['InvoiceDate'])

Currently, InvoiceDate is an object/string datatype which makes it difficult to extract individual date and time elements. Therefore, we convert it's datatype to datetime using pd.to_datetime function.

In [72]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 406829 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    406829 non-null  object        
 1   StockCode    406829 non-null  object        
 2   Description  406829 non-null  object        
 3   Quantity     406829 non-null  int64         
 4   InvoiceDate  406829 non-null  datetime64[ns]
 5   UnitPrice    406829 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      406829 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.9+ MB


Cleaned data is much more easier for analysis purpose and removing NaN CustomerID rows autometically removed NaN Description rows too.

In [73]:
df_clean.duplicated().sum()

np.int64(5225)

In [74]:
df_clean = df_clean.drop_duplicates()

Find and remove duplicate rows in dataset to avoid wrong analysis.

In [75]:
df_clean['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Greece', 'Singapore', 'Lebanon',
       'United Arab Emirates', 'Saudi Arabia', 'Czech Republic', 'Canada',
       'Unspecified', 'Brazil', 'USA', 'European Community', 'Bahrain',
       'Malta', 'RSA'], dtype=object)

Examining unique countries in the dataframe.

In [76]:
df_clean['TotalPrice'] = df_clean['Quantity']*df_clean['UnitPrice']

Make new column TotalPrice for analysis.

In [77]:
df_clean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom       15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom       22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom       20.34

In [78]:
df_clean = df_clean[df_clean['Quantity']>0]

Remove all columns with negative quantities which means returns.

In [79]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392732 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392732 non-null  object        
 1   StockCode    392732 non-null  object        
 2   Description  392732 non-null  object        
 3   Quantity     392732 non-null  int64         
 4   InvoiceDate  392732 non-null  datetime64[ns]
 5   UnitPrice    392732 non-null  float64       
 6   CustomerID   392732 non-null  float64       
 7   Country      392732 non-null  object        
 8   TotalPrice   392732 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 30.0+ MB


In [80]:
df_clean.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  392732.000000                         392732  392732.000000   
mean       13.153718  2011-07-10 19:15:24.576301568       3.125596   
min         1.000000            2010-12-01 08:26:00       0.000000   
25%         2.000000            2011-04-07 11:12:00       1.250000   
50%         6.000000            2011-07-31 12:02:00       1.950000   
75%        12.000000            2011-10-20 12:53:00       3.750000   
max     80995.000000            2011-12-09 12:50:00    8142.750000   
std       181.588420                            NaN      22.240725   

          CustomerID     TotalPrice  
count  392732.000000  392732.000000  
mean    15287.734822      22.629195  
min     12346.000000       0.000000  
25%     13955.000000       4.950000  
50%     15150.000000      12.390000  
75%     16791.000000      19.800000  
max     18287.000000  168469.600000  
std      1713.567773     311.083465

In [81]:
junkcodes = ['POST', 'M', 'DOT']
df_clean = df_clean[~df_clean['StockCode'].isin(junkcodes)]

Remove non product entries from anomalies analysis result.

In [82]:
unspec_count = df_clean[df_clean['Country']=='Unspecified'].shape[0]
print(f"Removing {unspec_count} rows with country 'Unspecified'")
df_clean = df_clean[df_clean['Country']!='Unspecified']

Removing 241 rows with country 'Unspecified'


Remove 241 unspecified countries entries from anomalies analysis result.

In [83]:
zero_count = df_clean[df_clean['UnitPrice']<0.5].shape[0]
print(f"Removing {zero_count} rows with 0 to 0.5 UnitPrice")
df_clean = df_clean[df_clean['UnitPrice']>=0.5]

Removing 38923 rows with 0 to 0.5 UnitPrice


Removed 38923 rows with 0 to 0.5 UnitPrice

In [84]:
df_clean.to_csv('cleaned_online_retail.csv',index=False)

Move cleaned dataframe to a csv file.